### ISSUE TITLE
Make the default volume configurable

### ISSUE THREAD

**Issue by mgoltzsche** [badaix/snapcast#910]
@badaix First of all thanks for providing and maintaining this awesome piece of software!

**Is your feature request related to a problem? Please describe.**
When connecting a new snapclient, its volume is set to 100%.
However this can be very annoying and even damage speakers when audio is already playing at that point.
To give an example I am [experimenting](https://github.com/mgoltzsche/highfipi) with a snapcast setup where a snapclient is run on a Raspberry Pi 3 with a Hifiberry Amp2 using the client option `--mixer hardware:Digital`. Setting the volume to 100% is very loud since it outputs 60W. Eventually it would even kill the old/cheap speakers I am using for testing purposes if I'd keep playing audio at that level (not sure if they got worse already after a couple of those accidents). Also I think my neighbors are angry with me now :wink:.

This relates to #735, particularly to https://github.com/badaix/snapcast/issues/735#issuecomment-757352479. cc @lucianm.

**Describe the solution you'd like**
To be safe, I'd like to be able to configure the default volume for new clients to eg. 30% to avoid such accidents.
I think it would be best to be able to make it configurable within the client. Though it would also be fine if it would be configurable on the server if that's easier to implement.

**Describe alternatives you've considered**
Stay focussed when connecting a new client, make sure no audio is playing, decrease the new client's volume before starting to play music.

**Comment by stuart12** [comment]
I have a similar use case for this feature request. My USB sound card does not have a hardware volume control so if snapserver  is sending audio when my machine starts, the audio plays at 100% until homeassistant starts and turns the volume in snapcast down. I'd thus like to set the default (or initial volume) to 0% or muted.

**Comment by nis65** [comment]
Is anyone working on that already? If not, I might have a look at the source if I would be able to implement this - at least with a little help from my friends :smile:. Actually I won't have time before February/March 2022, but I would love to have this feature...

**Comment by nis65** [comment]
@mgoltzsche @badaix I was thinking about where such a feature should be implemented and would be happy to receive your opinion:

From a users point of view I think it would be nice to set the initial snapclient volume as a snapclient parameter (so I could have individual defaults depending on whether I am deploying to a raspi with an AMP where the snapclient volume is the last volume control in the chain and a snapclient with a line out that is connected to an amp with hardware volume control.

But from a technical point of view the snapclient volume is set/defined on the server (is that correct?). So it is simpler (and also probably more suitable to the snapcast architecture) just to have a global "initial volume" that the snapserver assigns to each snapclient on initial connect.

**Comment by nis65** [comment]
Tried to understand the source / protocol
* Setting on the client side would mean I would have to extend the snapcast binary protocol (add a value to the hello message). this does not look like a job for me.
* Setting on server side seems to happen in one (or both) of the two places:
   * https://github.com/badaix/snapcast/blob/master/server/config.hpp#L61
   * https://github.com/badaix/snapcast/blob/master/server/config.hpp#L128
It's a very long time ago that I programmed c++, so I may be completely off... So next step would be to build up my compile/test pipeline...

**Comment by nis65** [comment]
Update: Finally had some the time to compile snapcast on my raspi. For a quick test, I manually replaced the snapserver binary in `/usr/bin/`with the one compiled by myself. Now the fun can begin...

**Comment by nis65** [comment]
And **YES**, I have a working solution :smile: 

I'll have to properly prepare a merge request now and then hope that it'll get merged sometime.

Have a preview [here](https://github.com/Daenou/ansible-multiroom-audio/issues/37#issuecomment-1166278020), you will have to compile from source to test it. I did it on a Raspberry4 (to prevent me from having to play with crosscompile), a full recompile of everything (you just need `snapserver` though) took about half an hour.

Please contact me directly if you want me to get a copy of my `snapserver` Raspberry binary.

**Comment by mgoltzsche** [comment]
Awesome! Looks good to me.

nis65 ***@***.***> schrieb am Sa., 25. Juni 2022, 15:02:

**Comment by nis65** [comment]
@badaix @mgoltzsche 

As the change is now merged into `develop`, what happens to the issue here? Will/can you @mgoltzsche close it (as you opened it) or will @badaix have to do this? 

And should this happen now or when the feature is promoted from `develop` to `master` only?

**Comment by badaix** [comment]
My usual process is to mark it as "next release", once it is in develop and to close it when it's merged into master and releases.
So I will not forget to mention it in the release notes and anyone wanting a similar feature will see it and know that it will be part of the next release

**Comment by nis65** [comment]
Perfect.So I'll await the next release 🙂 .

**Comment by badaix** [comment]
Fixed in [snapcast v0.27.0](https://github.com/badaix/snapcast/releases/tag/v0.27.0)
